lib/Sema/TypeCheckAccess.cpp

```diff
@@ -1421,43 +1426,103 @@ class ImplementationOnlyImportChecker
       const_cast<TypeRepr *>(typeRepr)->walk(TypeReprIdentFinder(
           [&](const ComponentIdentTypeRepr *component) {
         ModuleDecl *M = component->getBoundDecl()->getModuleContext();
-        if (!SF->isImportedImplementationOnly(M))
+        if (!SF.isImportedImplementationOnly(M))
           return true;
 
-        diagnose(component->getBoundDecl(), component);
+        diagnoseType(component->getBoundDecl(), component);
         foundAnyIssues = true;
         // We still continue even in the diagnostic case to report multiple
         // violations.
         return true;
       }));
     }
 
-    // FIXME: Are there places where we can be sure the TypeRepr fully accounts
-    // for the type, and therefore we could skip this extra work?
+    // Note that we can't skip checking the type even if the TypeRepr is okay,
+    // because that's how we check what conformances are being used.
+    //
+    // We still don't want to do this if we found issues with the TypeRepr,
+    // though, because that would result in some issues being reported twice.
     if (!foundAnyIssues && type) {
-      type.walk(SimpleTypeDeclFinder([&](const TypeDecl *typeDecl) {
-        ModuleDecl *M = typeDecl->getModuleContext();
-        if (!SF->isImportedImplementationOnly(M))
-          return TypeWalker::Action::Continue;
+      class ProblematicTypeFinder : public TypeDeclFinder {
+        const SourceFile &SF;
+        CheckImplementationOnlyTypeCallback diagnoseType;
+        CheckImplementationOnlyConformanceCallback diagnoseConformance;
+      public:
+        ProblematicTypeFinder(
+            const SourceFile &SF,
+            CheckImplementationOnlyTypeCallback diagnoseType,
+            CheckImplementationOnlyConformanceCallback diagnoseConformance)
+          : SF(SF), diagnoseType(diagnoseType),
+            diagnoseConformance(diagnoseConformance) {}
+
+        void visitTypeDecl(const TypeDecl *typeDecl) {
+          ModuleDecl *M = typeDecl->getModuleContext();
+          if (!SF.isImportedImplementationOnly(M))
+            return;
+
+          diagnoseType(typeDecl, /*typeRepr*/nullptr);
+        }
 
-        diagnose(typeDecl, /*typeRepr*/nullptr);
-        // We still continue even in the diagnostic case to report multiple
-        // violations.
-        return TypeWalker::Action::Continue;
-      }));
+        void visitSubstitutionMap(const SubstitutionMap &subs) {
+          for (ProtocolConformanceRef conformance : subs.getConformances()) {
+            if (!conformance.isConcrete())
+              continue;
+            const ProtocolConformance *concreteConf = conformance.getConcrete();
+
+            SubstitutionMap subConformanceSubs =
+                concreteConf->getSubstitutions(SF.getParentModule());
+            visitSubstitutionMap(subConformanceSubs);
```

Comment:
This recursive bit was tricky; it's to handle cases like

```swift
public struct ConditionalGenericStruct<T> {}
 extension ConditionalGenericStruct: NormalProto where T: NormalProto {
   public typealias Assoc = Int
 }
 public func testConditionalGeneric(_: NormalProtoAssocHolder<ConditionalGenericStruct<NormalStruct>>) {} // expected-error {{cannot use conformance of 'NormalStruct' to 'NormalProto' here; 'BADLibrary' has been imported as '@_implementationOnly'}}
```

Am I doing this right?

Reply:
I think you need to visit the associated conformances too; imagine you have:

```
protocol Q {}
protocol P {
  associatedtype T : Q
}
struct S1 : P {
  typealias T = S2
}
```

Let's say that S2 : Q is implementation only. I think your current check will miss that. Note that checking associated conformances should subsume your check.

Reply:
Also I'm worried about infinite recursion here. @DougGregor might have a better idea of the 'canonical' way to visit everything you need exactly once.

Reply:
I check associated conformances when doing conformance checking; this section is type signature checking. But I didn't think about infinite recursion in either case, so I'll need to pay more attention there.

Reply:
Okay, I *think* this is okay because substitution conformances always bottom out, even though associated conformances do not. And I don't need to check the associated conformances of my associated conformances because you can always check a conformance on its own while assuming the others are valid and then move on to the next one.